### ISSUE TITLE
Large memory usage by celery workers

### ISSUE THREAD

**Issue by paulpach** [WeblateOrg/weblate#2549]
I am seeing large memory usage in recent docker images,  which is causing my containers to get evicted.

```
top - 12:50:04 up 6 days, 12:50,  0 users,  load average: 0.06, 0.14, 0.20
Tasks:  22 total,   1 running,  21 sleeping,   0 stopped,   0 zombie
%Cpu(s):  4.7 us,  1.3 sy,  0.0 ni, 93.9 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem :  1735984 total,    84640 free,  1262892 used,   388452 buff/cache
KiB Swap:        0 total,        0 free,        0 used.   243032 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S %CPU %MEM     TIME+ COMMAND                                                                                          
    126 weblate   20   0  364492 119216  24048 S  0.0  6.9   0:03.05 celery                                                                                           
    127 weblate   20   0  364500 119160  23932 S  0.0  6.9   0:03.35 celery                                                                                           
    123 weblate   20   0  364488 119096  23904 S  0.0  6.9   0:03.06 celery                                                                                           
    125 weblate   20   0  368528 118460  25188 S  0.0  6.8   0:01.94 celery                                                                                           
    186 weblate   20   0  372788 116352  18008 S  1.3  6.7   0:04.02 uwsgi                                                                                            
    185 weblate   20   0  371252 114832  18008 S  0.0  6.6   0:02.50 uwsgi                                                                                            
    184 weblate   20   0  367708 110400  16552 S  0.0  6.4   0:00.45 uwsgi                                                                                            
    194 weblate   20   0  364244  99288   4516 S  0.0  5.7   0:00.00 celery                                                                                           
    195 weblate   20   0  364244  99288   4516 S  0.0  5.7   0:00.00 celery                                                                                           
    193 weblate   20   0  364244  99284   4516 S  0.0  5.7   0:00.00 celery                                                                                           
    190 weblate   20   0  364244  99280   4516 S  0.0  5.7   0:00.00 celery                                                                                           
    191 weblate   20   0  364236  99264   4508 S  0.0  5.7   0:00.00 celery                                                                                           
    192 weblate   20   0  364232  99264   4516 S  0.0  5.7   0:00.00 celery                                                                                           
    128 weblate   20   0  317044  84344  19524 S  0.0  4.9   0:01.76 uwsgi                                                                                            
    181 weblate   20   0  317044  68120   3300 S  0.0  3.9   0:00.00 uwsgi                                                                                            
    182 weblate   20   0  317044  68120   3300 S  0.0  3.9   0:00.00 uwsgi                                                                                            
    183 weblate   20   0  317044  68120   3300 S  0.0  3.9   0:00.00 uwsgi                                                                                            
      1 root      20   0   55380  20448   7096 S  0.0  1.2   0:00.40 supervisord                                                                                      
    124 root      20   0  140628  10164   8772 S  0.0  0.6   0:00.01 nginx                                                                                            
    129 www-data  20   0  140968   4264   2444 S  0.0  0.2   0:00.03 nginx                                                                                            
    211 root      20   0   38376   3372   2892 R  0.0  0.2   0:00.14 top                                                                                              
    206 root      20   0    4628    908    844 S  0.0  0.1   0:00.00 sh          
```

that is from a container that just started.    It looks like celery forks a lot and eats up a lot of memory.
Is it possible to switch celery to eventlet?

**Comment by nijel** [comment]
The problem is most likely caused by Whoosh, also see #2543 for related discussion.

I have no experience with eventlets in Celery, but I doubt it will make any difference here.

**Comment by paulpach** [comment]
I saw that bug, but decided to open a new one because I am experiencing it not only when it is doing search,  but it is constant.

eventlets could make a difference because celery won't fork 10 processes eating 100MB each.

**Comment by paulpach** [comment]
please see this PR:  https://github.com/WeblateOrg/docker/pull/133

**Comment by nijel** [comment]
I've played a bit with Celery today and there indeed seems to be something increasing memory usage. Until we have a proper solution, I've added configuration to restart worker if it consumes more than 200MB of memory in 0a3d373e2d88d73246d9e8f03334e4a3cde2d9a2 and https://github.com/WeblateOrg/docker/commit/3f9c0504c126cab2b24c574d04d51d34896193ec.

**Comment by SlavekB** [comment]
I can confirm – memory usage was increased during updating repository. And patch from commit 0a3d373 does not help on a large repository – see [comment on issue 2543](https://github.com/WeblateOrg/weblate/issues/2543#issuecomment-465311577).

**Comment by SlavekB** [comment]
@paulpach what version of Python do you use – 2.x or 3.x?

**Comment by paulpach** [comment]
@SlavekB I use the docker image,  I believe it has 3.x

**Comment by yajiedesign** [comment]
same issue.use docker image.

**Comment by nijel** [comment]
Apparently it needs more debugging and finding root cause.

**Comment by paulpach** [comment]
This is fixed for me with eventlet,   

I don't have an issue with updating

**Comment by nijel** [comment]
It's not really fixed, it's jest that you don't have x 100MB+ workers, but only one.

The on update memory consumption is probably from aggressive database caching we do to make updates fast and the cache can get quite big on big projects. The other problem is fulltext and translation memory updates (see #2543 ), those probably suffer from some Whoosh bugs and might not be fully addressed until we get rid of it (see #2825).

**Comment by paulpach** [comment]
@nijel this problem seems to be back,   looking at the docker image it seems the file moved to supervisord.conf?

**Comment by nijel** [comment]
Yes, we're now overriding whole supervisor configuration. This problem will be addressed in 4.0 by removing related code completely and relying on PostgreSQL instead, see https://github.com/WeblateOrg/weblate/issues/2825
